### ISSUE TITLE
network: Move to fully hotplugged network

### ISSUE THREAD

**Issue by sboeuf** [kata-containers/runtime#598]
We need to move to fully hotplugged network if we want to support the use case of booting a VM from a template, and also to allow more flexibility.

**Comment by sboeuf** [comment]
I wanted to talk about this issue that I had opened a few weeks ago because I am not so sure that's actually something we should do.

I got my PR about moving the OCI hooks out of virtcontainers, and not assuming for any default network being merged recently, and I was wondering if there was a benefit of moving to fully hotplugged model.
Indeed, as long as we make sure we're not making strong/wrong assumptions about adding the underlying network by default, implementation like `Frakti` can coexist and work with `kata-runtime`.

Also, I wanted to mention that moving to a fully hotplugged model might not be a good idea as some customer might want to be able to use some device that are not hotpluggable (passthrough of a NIC for instance).

@WeiZhang555 @bergwolf @caoruidong can you think of any strong reason why we should move to a fully hotplugged model for networking?

/cc @egernst @amshinde

**Comment by bergwolf** [comment]
@sboeuf I have already done the change (see https://github.com/bergwolf/kata-runtime/tree/network) but there are issues w/ network hotplug[1] so I didn't raise any PR yet. My strong reason for the move is that vm factory requires it otherwise different VMs created from a template will use the same nic config (mostly the mac address). Another (less-strong) reason is to support a use case that a container requires **NO** external network connections in which case we should not give it a nic at all.

[1]. As I was testing the full network hotplug model, I found that the agent uevent listening code not to be working as expected. If I hotplug a nic, no matter how long I wait in the `UpdateInterface` call, it never gets the uevents, not until the RPC returns. It feels like agent is single threaded that cannot bring up the uevent listening goroutine while waiting in the `UpdateInterface` call. @sboeuf @caoruidong how did you guys test network hotplug when implementing it?

And I see that the block device hotplug code actually waits and listens on the uevents on its own. Does it mean the agent uevent listening goroutine does not work for block device hotplug either? @devimc @amshinde

**Comment by sboeuf** [comment]
@bergwolf 
>My strong reason for the move is that vm factory requires it otherwise different VMs created from a template will use the same nic config (mostly the mac address).

The way things are implemented now, as long as you have a factory, no network will be added by default, am I missing something?

>Another (less-strong) reason is to support a use case that a container requires NO external network connections in which case we should not give it a nic at all.

This one already works I think. If no network is provided by the upper layer, the network namespace will be empty and nothing will be found, hence nothing will be added to the VM.
Also, we could think about adding a clear boolean flag to the SandboxConfig, instead of relying on the existence of a factory, in order to decide if we want the network scanning to be performed or not.

>As I was testing the full network hotplug model, I found that the agent uevent listening code not to be working as expected. If I hotplug a nic, no matter how long I wait in the UpdateInterface call, it never gets the uevents, not until the RPC returns. It feels like agent is single threaded that cannot bring up the uevent listening goroutine while waiting in the UpdateInterface call. @sboeuf @caoruidong how did you guys test network hotplug when implementing it?

That's interesting as some time ago, I had some issues with the uevent listener as it was not working as expected for me. We might need to narrow this one down since we want to make sure using the uevent listener is a reliable way of detecting new interfaces.
About the way I have tested the network hotplug, I have mainly tested with the network monitor process, and also, I had spent some time testing with the full network hotplug PR (but it's not usable right now as not rebased).

**Comment by bergwolf** [comment]
@sboeuf 

> The way things are implemented now, as long as you have a factory, no network will be added by default, am I missing something?

While it is true, it is also the reason vm factory is not fully functional. We need a fully hotplug model to have network setup at least for vm factory case. But again, do we want to diverge that much?

> This one already works I think. If no network is provided by the upper layer, the network namespace will be empty and nothing will be found, hence nothing will be added to the VM.

It is working for vm factory, and working for non-factory at the API level. We surely need to change CLI code to make it work in a generic case. But that is just a different approach. If we have fully hotplug network model, we gain the functionality naturally.

>I had some issues with the uevent listener as it was not working as expected for me. 

Yes, it is exactly the issue I was hitting. I guess for the same reason @amshinde added waiting in block device hotplug code on its own.

**Comment by caoruidong** [comment]
I occasionally meet the issue, not very often. Will https://github.com/kata-containers/agent/pull/306 help?

**Comment by bergwolf** [comment]
@caoruidong Nop. I am constantly seeing the failure with kata-agent master.

**Comment by sboeuf** [comment]
@bergwolf 
As long as we can cover all the use cases our user could have, I think we should keep the coldplug as a backup for some specific use cases.
By __stong reason__, I was asking if there is one use case that cannot be covered by the fact that we're not using fully hotplugged network, but I don't think we have such case.
Also, for the record, I thought that moving to a fully hotplugged network was a good idea because I was interested in reducing the code complexity, but @egernst mentioned that some users might use Kata with Docker and that hotplugging the network, which means supporting network hotplug with the hypervisor, might not be strictly needed. Let's try not to force our user to support some hotplug features (from an hypervisor perspective), when the use case does not require it.

@amshinde could you take a look at this sporadic issue with uevent?

**Comment by WeiZhang555** [comment]
> As long as we can cover all the use cases our user could have, I think we should keep the coldplug as a backup for some specific use cases.

I don't have a **strong reason**, so I think we can choose to support hotplug whenever it can, and leave codeplug as a backup. This is fine for me.

**Comment by bergwolf** [comment]
@sboeuf @WeiZhang555 Sure, I understand. I just don't want to diverge too much between vm factory and non-factory code path. I am fine with keeping coldplug.

So the question is, do we determine the hotplug/codeplug switch depending on hypervisor capability (that's what I'd call a _backup plan_)? Or do we always code plug in non vm factory case?

**Comment by WeiZhang555** [comment]
> So the question is, do we determine the hotplug/codeplug switch depending on hypervisor capability (that's what I'd call a backup plan)? Or do we always code plug in non vm factory case?

Vote to "depending on hypervisor capability", and always use hotplug unless the hypervisor can't support.

**Comment by sboeuf** [comment]
@bergwolf @WeiZhang555 
By __depending on hypervisor capability__, do you mean actively testing the hypervisor, or simply add a new field in the `configuration.toml` under the `hypervisor` section?
I think the configuration option would be the most reliable one.

Now, that being said, I don't know why we should try to hotplug by default. Does it buy us some time? Do we increase our performances? If not, I don't see why we should bother changing this.

**Comment by sboeuf** [comment]
@WeiZhang555 
> it can deduplicate codes then increase the readability and make it easier to maintain.

Unfortunately, moving to full hotplug will not allow to support use case where the hypervisor does not support hotplug.

@bergwolf 
>It's more about the default behavior you are asking. I am fine if you insist on coldplug it, provided that the network hotplug code path is also extensively tested since that is a feature we must support in all kinds of scenarios (k8s cni, docker connect etc.). I'm worried about network hotplug stability because just recently I found out that it is not working and clearly we do not have any coverage for it in CI.

About CI and making sure network hotplug works, I cannot agree more :)
We obviously need the network hotplug to work since we definitely need this capability!
